File: modules/protocol/769/src/main/java/candle/protocol/codec/VarIntCodec.java
package candle.protocol.codec;

import candle.types.primitives.Int32;
import candle.types.resource.Identifier;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class VarIntCodec extends TypeCodec<Int32> {
  private static TypeIdentifier<Int32> IDENTIFIER;

  public VarIntCodec() {
    if ( IDENTIFIER == null ) {
      IDENTIFIER = identifier("minecraft", "string");
    }
  }

  public static TypeIdentifier<Int32> identifier() {
    return IDENTIFIER;
  }

  @Override
  public Int32 read( InputStream in, AbstractDataCodec dataCodec ) throws
                                                                   IOException {
    int numRead = 0;
    int result = 0;
    byte read;
    do {
      int byteVal = in.read();
      if ( byteVal == -1 ) {
        throw new IOException("Stream ended while reading VarInt");
      }
      read = (byte) byteVal;
      result |= ( read & 0x7F ) << ( 7 * numRead );
      numRead++;
      if ( numRead > 5 ) {
        throw new IOException("VarInt too long");
      }
    } while ( ( read & 0x80 ) != 0 );
    return new Int32(result);
  }

  @Override
  public void write( OutputStream out, Int32 value, AbstractDataCodec dataCodec ) throws
                                                                                  IOException {
    int intValue = value.value();
    while ( ( intValue & ~0x7F ) != 0 ) {
      out.write(( intValue & 0x7F ) | 0x80);
      intValue >>>= 7;
    }
    out.write(intValue);
  }

  @Override
  public Identifier namespacedIdentifier() {
    return IDENTIFIER;
  }
}

File: protocol/769/src/main/kotlin/net/candlemc/protocol/codec/VarIntCodec.kt
package net.candlemc.protocol.codec

import net.candlemc.types.resource.Identifier
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

class VarIntCodec : TypeCodec<Int>() {
    init {
        if (IDENTIFIER == null) {
            IDENTIFIER = identifier("minecraft", "var_int")
        }
    }

    @Throws(IOException::class)
    override fun read(inputStream: InputStream, dataCodec: AbstractDataCodec): Int {
        var numRead = 0
        var result = 0
        var read: Byte
        do {
            val byteVal = inputStream.read()
            if (byteVal == -1) {
                throw IOException("Stream ended while reading VarInt")
            }
            read = byteVal.toByte()
            result = result or ((read.toInt() and 0x7F) shl (7 * numRead))
            numRead++
            if (numRead > 5) {
                throw IOException("VarInt too long")
            }
        } while ((read.toInt() and 0x80) != 0)
        return result
    }

    @Throws(IOException::class)
    override fun write(outputStream: OutputStream, value: Int, dataCodec: AbstractDataCodec) {
        var intValue: Int = value
        while ((intValue and 0x7F.inv()) != 0) {
            outputStream.write((intValue and 0x7F) or 0x80)
            intValue = intValue ushr 7
        }
        outputStream.write(intValue)
    }

    override fun namespacedIdentifier(): Identifier {
        return IDENTIFIER!!
    }

    companion object {
        private var IDENTIFIER: TypeIdentifier<Int>? = null

        fun identifier(): TypeIdentifier<Int> {
            if (IDENTIFIER == null) throw IllegalStateException("VarIntCodec was not initialized")
            return IDENTIFIER!!
        }
    }
}
